pkg/modifycontroller/controller.go

```diff
@@ -0,0 +1,378 @@
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+)
+
+// ModifyController watches PVCs and checks if they are requesting an modify operation.
+// If requested, it will modify the volume according to parameters in VolumeAttributesClass
+type ModifyController interface {
+	// Run starts the controller.
+	Run(workers int, ctx context.Context)
+}
+
+type modifyController struct {
+	name            string
+	modifier        resizer.Resizer
+	kubeClient      kubernetes.Interface
+	claimQueue      workqueue.RateLimitingInterface
+	eventRecorder   record.EventRecorder
+	pvSynced        cache.InformerSynced
+	pvcSynced       cache.InformerSynced
+	vacSynced       cache.InformerSynced
+	podLister       corelisters.PodLister
+	podListerSynced cache.InformerSynced
+
+	// a cache to store PersistentVolume objects
+	volumes cache.Store
+	// a cache to store PersistentVolumeClaim objects
+	claims cache.Store
+	// a cache to store VolumeAttributesClass objects
+	volumeAttributesClasses cache.Store
+	// a cache to store modify volume in progress or infeasible PVCs in uncertain state
+	uncertainPVCs map[string]v1.PersistentVolumeClaim
+}
+
+// NewModifyController returns a ModifyController.
+func NewModifyController(
+	name string,
+	modifier resizer.Resizer,
+	kubeClient kubernetes.Interface,
+	resyncPeriod time.Duration,
+	informerFactory informers.SharedInformerFactory,
+	pvcRateLimiter workqueue.RateLimiter) ModifyController {
+	pvInformer := informerFactory.Core().V1().PersistentVolumes()
+	pvcInformer := informerFactory.Core().V1().PersistentVolumeClaims()
+	vacInformer := informerFactory.Storage().V1alpha1().VolumeAttributesClasses()
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartStructuredLogging(0)
+	eventBroadcaster.StartRecordingToSink(&corev1.EventSinkImpl{Interface: kubeClient.CoreV1().Events(v1.NamespaceAll)})
+	eventRecorder := eventBroadcaster.NewRecorder(scheme.Scheme,
+		v1.EventSource{Component: fmt.Sprintf("external-resizer %s", name)})
+
+	claimQueue := workqueue.NewNamedRateLimitingQueue(
+		pvcRateLimiter, fmt.Sprintf("%s-pvc", name))
+
+	ctrl := &modifyController{
+		name:                    name,
+		modifier:                modifier,
+		kubeClient:              kubeClient,
+		pvSynced:                pvInformer.Informer().HasSynced,
+		pvcSynced:               pvcInformer.Informer().HasSynced,
+		vacSynced:               vacInformer.Informer().HasSynced,
+		claimQueue:              claimQueue,
+		volumes:                 pvInformer.Informer().GetStore(),
+		claims:                  pvcInformer.Informer().GetStore(),
+		volumeAttributesClasses: vacInformer.Informer().GetStore(),
```

Comment:
Which lister are you referring to? There are some built in functions in the Store we are using as well.

Reply:
because lister is more general way to fetch objects from inform store. you can see that listers are used a lot in the kube-controller-manager、kubelet or some operators.

https://github.com/sunnylovestiramisu/external-resizer/blob/9c91f9e817dfc92c1b998173d31cb8337fad657f/vendor/k8s.io/client-go/informers/storage/v1alpha1/volumeattributesclass.go#L39

Reply:
@gnufied Why did you use the informer for resizer not the lister?

Reply:
We are using this: ctrl.volumes.GetByKey(oldPVC.Spec.VolumeName).

So does the lister has something equivalent?

Reply:
But if we are only reading object from informer,  we should be okay no?

Reply:
Yes, lister can get or list objects

Reply:
I don't think as such we need a `volumeAttributesClasses` store, because as I mentioned before - part of the reason for using store directly here was to avoid waiting for informers to be repopulated. This is a design, somewhat borrowed from pv-controller but our understanding of these practices evolve. 

In this particular case, `volumeAttributeClasses` should not even change all that much, so we definitely don't need a `cache.Store` for that object.


Similarly, I don't think we have to copy the choice of resize controller updating pvc and pv objects in `cache.Store` after updating those objects in api-server. I would like to change this in resize controller too, but before I do that - I would like to check how the controller behaves. But we don't have to copy that choice in this controller. cc @msau42

Reply:
Then we cannot use the same method function for these two either. Will keep them separated for now.

Reply:
It does not change means we should use the lister for VAC? And we expect the lister to get updated.

Right now we use it for check existence:

```
vacObj, exists, err := ctrl.volumeAttributesClasses.GetByKey(*pvcSpecVacName)
````

Reply:
We can keep a lister and `hasSynced` methods. We shouldn't have to access the store directly.